source/extensions/io_socket/buffered_io_socket/peer_buffer.h

```diff
@@ -0,0 +1,79 @@
+#pragma once
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/common/pure.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace IoSocket {
+namespace BufferedIoSocket {
+
+/**
+ * The interface for the writer.
+ */
+class WritablePeer {
+public:
+  virtual ~WritablePeer() = default;
+
+  /**
+   * Set the flag to indicate no further write from peer.
+   */
+  virtual void setWriteEnd() PURE;
+  /**
+   * @return true if the peer promise no more write.
+   */
+  virtual bool isWriteEndSet() PURE;
+
+  /**
+   * Raised when peer is destroyed. No further write to peer is allowed.
+   */
+  virtual void onPeerDestroy() PURE;
+
+  /**
+   * Notify that consumable data arrives. The consumable data can be either data to read, or the end
+   * of stream event.
+   */
+  virtual void maybeSetNewData() PURE;
+
+  /**
+   * @return the buffer to be written.
+   */
+  virtual Buffer::Instance* getWriteBuffer() PURE;
+
+  /**
+   * @return false more data is acceptable.
+   */
+  virtual bool isWritable() const PURE;
+
+  /**
+   * @return true if peer is valid and writable.
+   */
+  virtual bool isPeerWritable() const PURE;
+
+  /**
+   * Raised by the peer when the peer switch from high water mark to low.
+   */
+  virtual void onPeerBufferWritable() PURE;
+};
+
+/**
+ * The interface as the union of ReadableSource and WritablePeer.
+ */
+class ReadWritable : public WritablePeer {
```

Comment:
> Could we have separate interfaces or a single interface class that provides everything instead of having the interfaces inherit from each other?

In the previous iteration, I removed the `Readable` interface because `Readable` is never used independently according to your comment.

Alternatively, I can bind a WritablePeer and Readable to two members of user_file_event_, pretending that they are independent members. I don't see the value.

Reply:
Rename `ReadWritable` to per your `UserspaceIoHandle` comment https://github.com/envoyproxy/envoy/pull/13418#discussion_r520944878